Avoid forwarding dns queries of type DS to recursors (#6183)

vierbergenlars opened the pull request:
Answer all `DS` queries that are sent for our domains to recursors with a refused response code.

The upstream servers will not have a `DS` record for consul domains anyways, because Consul does not support DNSSEC yet, but also because the domain will not be known by the resolvers.

Fixes #6183

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/consul?pullRequest=6207)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/consul?pullRequest=6207)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/consul?pullRequest=6207).</sub>

i0rek commented:
Thank you @vierbergenlars for your work. Is this still something you would like to add?

vierbergenlars commented:
Yes, this is still something I would like to add

i0rek commented:
Thanks for your reply! This sounds like a nice patch and I would like to ask you to rebase it so that I can review.

vierbergenlars commented:
Rebased on top of master.

Note that the diff is best viewed with whitespace changed disabled, because I added a condition and had to indent a large block of code.

i0rek commented on agent/dns.go:
```diff
@@ -1760,56 +1760,79 @@ func (d *DNSServer) handleRecurse(resp dns.ResponseWriter, req *dns.Msg) {
-			return
+	// Refuse to recursively resolve requests of type DS for domains that we manage.
+	// Upstream resolvers will not know about the correct record anyways, and it might result in a resolver loop
+	if q.Qtype == dns.TypeDS {
```
Could you extract a func for this block? Eg `func (d *DNSServer) refuseRecurseDS(req *dns.Msg, domain string) bool`.

And also a func for creating the message `func generateMessage(code int)`.

Afterwards we can write:

```golang
for _, domain := range []string{d.domain, d.altDomain} {
  if d.refuseRecurseDS(req, domain) {
    d.logger.Debug("request for %v (%s): Refusing to recurse type DS requests to %v.", q, network, domain)
    m := generateMessage(dns.RcodeRefused)
    resp.WriteMsg(m)
    return
  }
}
```

That would also get rid of the big indentation below. And `generateMessage` could be reused for the `dns.RcodeServerFailure` message at the end.
